controller/src/lib.rs

```diff
@@ -254,13 +335,23 @@ impl Controller {
             message,
             data: None,
         };
-        let reply = self.rpc(request).await.unwrap();
+        let reply = self.rpc(request).await?;
+
+        // If we get back a ReadFailed error, one possibility is that we asked
+        // to read a transceiver that's not present.
+        let data = match reply.message.body {
+            MessageBody::SpResponse(SpResponse::Error(MessageError::ReadFailed)) => {
+                return Err(Error::ReadFailed);
+            }
+            MessageBody::SpResponse(SpResponse::Error(e)) => return Err(Error::from(e)),
+            MessageBody::SpResponse(SpResponse::Read(_)) => reply.data.unwrap(),
+            _ => panic!("Unexpected message type"),
```

Comment:
This might be okay, but it's not obvious that it is - does `self.rpc()` guarantee that `reply.message.body` will fall into one of the three cases above? Could this return something like `Error::UnexpectedMessageType(maybe_some_details_maybe_not)` instead of panicking?

Reply:
I think I meant to write `todo!()` here rather than `panic!()`! Still, I think we could return an error here. `self.rpc()` guarantees that we get a `MessageBody::SpResponse(_)`, but not what that inner variant is. I should return a protocol error or similar, you're right. Thanks!